packages/flutter_tools/bin/fuchsia_asset_builder.dart

```diff
@@ -0,0 +1,110 @@
+const String _kOptionManifest = 'manifest';
+const String _kOptionAssetManifestOut = 'asset-manifest-out';
+const List<String> _kRequiredOptions = const <String>[
+  _kOptionPackages,
+  _kOptionWorking,
+  _kOptionAssetManifestOut,
+];
+
+Future<Null> main(List<String> args) async {
+  final AppContext executableContext = new AppContext();
+  executableContext.setVariable(Logger, new StdoutLogger());
+  await executableContext.runInZone(() {
+    // Initialize the context with some defaults.
+    // This list must be kept in sync with lib/executable.dart.
+    context.putIfAbsent(Stdio, () => const Stdio());
+    context.putIfAbsent(Platform, () => const LocalPlatform());
+    context.putIfAbsent(FileSystem, () => const LocalFileSystem());
+    context.putIfAbsent(ProcessManager, () => const LocalProcessManager());
+    context.putIfAbsent(Logger, () => new StdoutLogger());
+    context.putIfAbsent(Cache, () => new Cache());
+    context.putIfAbsent(Config, () => new Config());
+    context.putIfAbsent(OperatingSystemUtils, () => new OperatingSystemUtils());
+    context.putIfAbsent(Usage, () => new DisabledUsage());
+    return run(args);
+  });
+}
+
+Future<Null> writeFile(File outputFile, DevFSContent content) async {
+  outputFile.createSync(recursive: true);
+  final List<int> data = await content.contentsAsBytes();
+  outputFile.writeAsBytesSync(data);
+  return null;
+}
+
+Future<Null> run(List<String> args) async {
+  final ArgParser parser = new ArgParser()
+    ..addOption(_kOptionPackages, help: 'The .packages file')
+    ..addOption(_kOptionWorking,
+        help: 'The directory where to put temporary files')
+    ..addOption(_kOptionManifest, help: 'The manifest file')
+    ..addOption(_kOptionAssetManifestOut);
+  final ArgResults argResults = parser.parse(args);
+  if (_kRequiredOptions
+      .any((String option) => !argResults.options.contains(option))) {
+    printError('Missing option! All options must be specified.');
+    exit(1);
+  }
+  Cache.flutterRoot = platform.environment['FLUTTER_ROOT'];
+
+  final String workingDir = argResults[_kOptionWorking];
+  final AssetBundle assets = await buildAssets(
+    manifestPath: argResults[_kOptionManifest] ?? defaultManifestPath,
+    workingDirPath: workingDir,
+    packagesPath: argResults[_kOptionPackages],
+    includeDefaultFonts: false,
+  );
+
+  if (assets == null) {
+    print('Unable to find assets.');
+    exit(1);
+  }
+
+  final List<Future<Null>> calls = <Future<Null>>[];
+  assets.entries.forEach((String fileName, DevFSContent content) {
+    final File outputFile = fs.file(fs.path.join(workingDir, fileName));
+    calls.add(writeFile(outputFile, content));
+  });
+  await Future.wait(calls);
+
+  final String outputMan = argResults[_kOptionAssetManifestOut];
+  writeFuchsiaManifest(assets, argResults[_kOptionWorking], outputMan);
+  exit(0);
+}
+
+void writeFuchsiaManifest(AssetBundle assets, String outputBase, String fileDest) {
```

Comment:
May as well make this async so you can avoid the sync file I/O

Reply:
The synchronous file I/O was on purpose. I'm not sure of the guarantees Dart provides, but I wanted to make sure the files would be written before the program terminates. I also wanted to make sure that the file writes occur in order. Can you help me understand if my concerns are handled by Dart and I should just make this async?

Reply:
You can just use `await` on the async calls to provide all those guarantees.

Reply:
In this case don't I end up awaiting all the async calls in this function and then the function itself in the call to it?

Reply:
Yep.  It'd be:

```
  ...
  await writeFuchsiaManifest(assets, argResults[_kOptionWorking], outputMan);
}

Future<Null> writeFuchsiaManifest(AssetBundle assets, String outputBase, String fileDest) async {
  syncOp();
  await asyncOp();
  ...
}
```

The benefits are:
* It frees up the main isolate to do other work in the meantime (e.g. animating a "waiting..." cursor if you've set one up).
* It allows you to add a timeout that will cause the tool to fail rather than just hanging forever.  In the main flutter tool, we've found places where file I/O backed by a network file system can hang, and if you're using synchronous operations, your whole tool just hangs.

Reply:
Ah, okay, thanks for explaining why this is worth doing. I don't really write Dart code so I'm not as aware of these implications as I maybe should be.

Reply:
Done